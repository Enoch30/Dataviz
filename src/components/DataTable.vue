<template>
  <v-data-table
    :headers="headers"
    :items="data"
    sort-by="calories"
    class="elevation-1"
  >
    <template v-slot:top>
      <v-toolbar flat color="white">
        <v-toolbar-title>Edit Revenues</v-toolbar-title>
        <v-divider
          class="mx-4"
          inset
          vertical
        ></v-divider>
        <v-spacer></v-spacer>
        <v-dialog v-model="dialog" max-width="500px">
          <template v-slot:activator="{ on }">
            <v-btn color="indigo" name='add-new-btn' dark class="mb-2" v-on="on">Add New</v-btn>
          </template>
          <v-card>
            <v-card-title>
              <span class="headline">{{ 'formTitle' }}</span>
            </v-card-title>

            <v-card-text>
              <v-container>
                <v-row>
                  <v-col cols="12" sm="6" md="4">
                    <v-text-field v-model="editedItem.year" label="Year" name='year'></v-text-field>
                  </v-col>
                  <v-col cols="12" sm="6" md="4">
                    <v-text-field v-model="editedItem.sale" label="Sale" name='sale'></v-text-field>
                  </v-col>
                  <v-col cols="12" sm="6" md="4">
                    <v-text-field v-model="editedItem.transactions" label="Transactions" name='transactions'></v-text-field>
                  </v-col>
                </v-row>
              </v-container>
            </v-card-text>

            <v-card-actions>
              <v-spacer></v-spacer>
              <v-btn color="blue darken-1" text @click="close">Cancel</v-btn>
              <v-btn id='save-btn' color="blue darken-1" text @click="save">Save</v-btn>
            </v-card-actions>
          </v-card>
        </v-dialog>
      </v-toolbar>
    </template>
    <template v-slot:item.action="{ item }">
      <v-icon
       small
        class="m-2"
        @click="editItem(item)"
        id=edit
      >{{ icons.mdiPencil }}
      </v-icon>

      <v-icon
       small
        class="m-2"
        id=delete
        @click="deleteItem(item)"
      >{{ icons.mdiDelete }}
      </v-icon>

    </template>
    <template v-slot:no-data>
      <v-btn color="primary" @click="initialize">Reset</v-btn>
    </template>
  </v-data-table>
  
</template>

<script>
  import {
      mdiAccount,
      mdiPencil,
      mdiShareVariant,
      mdiDelete,
    } from '@mdi/js';

  export default {
    props: ['salesData'],
    data: () => ({
      dialog: false,
      headers: [
        {
          text: 'Year',
          align: 'left',
          sortable: false,
          value: 'year',
        },
        { text: 'Sale', value: 'sale' },
        { text: 'Transactions', value: 'transactions' },
        { text: 'Actions', value: 'action', sortable: false },
      ],
      data: [],
      //Stores information on new table entries
      editedIndex: -1,
      editedItem: {
        year: '',
        sale: '',
        transactions: '',
      },
      //initial values before user updates fields
      defaultItem: {
        year: '',
        sale: '',
        transactions: '',
      },
      icons: {
        mdiAccount,
        mdiPencil,
        mdiShareVariant,
        mdiDelete,
      },
    }),

    computed: {
      formTitle () {
        //Reflects form name with currnt action
        return this.editedIndex === -1 ? 'New Item' : 'Edit Item'
      },
    },

    watch: {
      dialog (val) {
        val || this.close()
      },
    },

    created () {
      this.initialize()
    },

    methods: {
      //extract relevant data from the received object
      initialize () {
        var i = 0
        for ( i = 0; i < this.salesData.labels.length; i++){
          this.data.push({
            year: this.salesData.labels[i],
            sale: this.salesData.datasets[0].data[i],
            transactions: this.salesData.transactions[i]
          })
        }
      },

      editItem (item) {
        this.editedIndex = this.data.indexOf(item)
        this.editedItem = Object.assign({}, item)
        this.dialog = true
      },

      deleteItem (item) {
        const index = this.data.indexOf(item)
        confirm('Are you sure you want to delete this item?') && this.data.splice(index, 1)
      },

      close () {
        this.dialog = false
        setTimeout(() => {
          this.editedItem = Object.assign({}, this.defaultItem)
          this.editedIndex = -1
        }, 300)
      },

      save () {
        if (this.editedIndex > -1) {
          Object.assign(this.data[this.editedIndex], this.editedItem)
        } else {
          this.data.push(this.editedItem)
        }
        this.close()
      },
    },
  }
</script>