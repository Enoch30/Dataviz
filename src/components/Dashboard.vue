<template>
  <div>
    <mdb-container class='max-width hidden-sm-and-down'>
      
      <!-- 
        Removed due to Google Maps API key
        <GChart
        :settings="{ packages: ['corechart', 'table', 'map'] }"
        type="Map"
        :data="[
          salesData.labels,
          salesData.datasets.data
        ]
        "
        :options="chartOptions"
      /> -->
      <mdb-row>
        <mdb-col md="6" >
          <template>
            <mdb-card class="chart">
              <mdb-card-header color="indigo lighten-1">Pie chart</mdb-card-header>
              <mdb-card-body>
              <mdb-card-title></mdb-card-title>
              <div class="d-flex justify-content-center">
                  <mdb-pie-chart
                datalabels
                :data="salesData"
                :options="pieChartOptions"
                :width="470"
              />
            </div>
            </mdb-card-body>
            </mdb-card>
          </template>
        </mdb-col>

        <mdb-col md="6">
          <template>
            <mdb-card class="chart">
              <mdb-card-header color="indigo lighten-1">Bar chart</mdb-card-header>
              <mdb-card-body>
              <mdb-card-title></mdb-card-title>
              <div class="d-flex justify-content-center">
                <mdb-bar-chart 
                  :data="salesData"
                  :options="barChartOptions"
                  :width="470"
                  >
                </mdb-bar-chart>
            </div>
            </mdb-card-body>
            </mdb-card>
          </template>
        </mdb-col>     
      </mdb-row>
    </mdb-container>
     <mdb-card class="m-3">
        <mdb-card-header color="indigo lighten-1">Data Table</mdb-card-header>
        <!--Sales Table-->
        <DataTable :salesData='salesData' />
     </mdb-card>
  </div>
</template>

<script>
import { mdbPieChart, mdbContainer, mdbBarChart } from "mdbvue";
import DataTable from '../components/DataTable';
import {mdbRow, mdbCol} from 'mdbvue';
 import { mdbCard, mdbCardBody, mdbCardHeader, mdbCardTitle } from 'mdbvue';
//  import { GChart } from 'vue-google-charts';

  export default {
    name: "Dashboard",
    components: {
      DataTable,
      mdbBarChart,
      mdbPieChart,
      mdbContainer,
      mdbRow,
      mdbCol,
      mdbCard,
      mdbCardBody,
      mdbCardHeader,
      mdbCardTitle,
      // GChart
    },
    data() {
      return {
        //Primary Data
        salesData: {
          labels: ["2015", "2016", "2017", "2018", "2019"],
          transactions: [5000, 3200, 1238, 2834, 2672],
          datasets: [
            {
              label: "Sales per year",
              data: [300, 50, 100, 40, 120],
              backgroundColor: [
                "#F7464A",
                "#46BFBD",
                "#FDB45C",
                "#949FB1",
                "#4D5360"
              ],
              hoverBackgroundColor: [
                "#FF5A5E",
                "#5AD3D1",
                "#FFC870",
                "#A8B3C5",
                "#616774"
              ],
              borderWidth: 1
            }
          ]
        },

        //Pie chart settings
        pieChartOptions: {
          responsive: true,
          maintainAspectRatio: false,
          plugins: {
            datalabels: {
              color: "white",
              align: "center",
              font: {
                size: 16
              },
              //formates the table values
              formatter: value => {
                const [dataset] = this.salesData.datasets;
                const setValue = dataset.data.reduce((a, b) => a + b);
                return `${Math.round((value / setValue) * 100)}%`;
              }
            }
          }
        },
      
      //Bar chart settings
      barChartOptions: {
          responsive: true,
          maintainAspectRatio: false,
          scales: {
            xAxes: [
              {
                barPercentage: 1,
                gridLines: {
                  display: true,
                  color: "rgba(0, 0, 0, 0.1)"
                }
              }
            ],
            yAxes: [
              {
                gridLines: {
                  display: true,
                  color: "rgba(0, 0, 0, 0.1)"
                }
              }
            ]
          }
        }
      };
    }
  }

    
</script>
<style scoped>
h3 {
  font-weight: normal;
  padding-top: 20px;
  padding-bottom: 30px;
}
p {
  color: #969696;
  margin-bottom: 0;
  font-size: 14px;
}
</style>
