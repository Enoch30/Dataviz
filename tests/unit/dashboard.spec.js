import { shallowMount, createLocalVue } from '@vue/test-utils'
import Dashboard from "@/components/Dashboard.vue"
import DataTable from '@/components/DataTable.vue'
import vuetify from 'vuetify'

let testData = {
    labels: ["2015", "2016"],
    transactions: [5000, 3200],
    datasets: [
    {
        label: "Sales per year",
        data: [300, 50],
        backgroundColor: [
        "#F7464A",
        "#46BFBD",
        ],
        hoverBackgroundColor: [
        "#FF5A5E",
        "#616774"
        ],
    }
  ]
}

//Basic tests, see e2e tests
describe('Dashboard.vue', () => {
    let wrapper;
    let title = 'Dataviz'
    beforeEach(()=>{
        const localVue = createLocalVue()
        localVue.use(vuetify)
        wrapper = shallowMount(Dashboard, {
          localVue,
          propsData: { 
            salesData: testData,
          }
        });
      })

    it('renders a vue instance', () => {
        expect(shallowMount(Dashboard).isVueInstance()).toBe(true);
    });

    it('can mount table', () => {
        const wrapper = shallowMount(Dashboard, {
            propsData: { 
                salesData: testData
            }
          })
        expect(wrapper.find(DataTable).exists()).toBe(true)
    })
})



