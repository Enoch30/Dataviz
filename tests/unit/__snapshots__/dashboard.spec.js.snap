// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard.vue should have a custom title and match snapshot 1`] = `
<div title="Foobar">
  <mdb-container-stub tag="div" class="max-width hidden-sm-and-down">
    <v-btn color="blue darken-1" id="btn" text="" class="btn">Cancel</v-btn>
    <mdb-row-stub tag="div">
      <mdb-col-stub tag="div" md="6">
        <mdb-card-stub tag="div" class="chart">
          <mdb-card-header-stub tag="div" color="indigo lighten-1">Pie chart</mdb-card-header-stub>
          <mdb-card-body-stub tag="div">
            <mdb-card-title-stub tag="h4"></mdb-card-title-stub>
            <div class="d-flex justify-content-center">
              <mdb-pie-chart-stub chartid="pie-chart" width="470" height="400" cssclasses="" plugins="" data="[object Object]" options="[object Object]" datalabels="true"></mdb-pie-chart-stub>
            </div>
          </mdb-card-body-stub>
        </mdb-card-stub>
      </mdb-col-stub>
      <mdb-col-stub tag="div" md="6">
        <mdb-card-stub tag="div" class="chart">
          <mdb-card-header-stub tag="div" color="indigo lighten-1">Bar chart</mdb-card-header-stub>
          <mdb-card-body-stub tag="div">
            <mdb-card-title-stub tag="h4"></mdb-card-title-stub>
            <div class="d-flex justify-content-center">
              <mdb-bar-chart-stub chartid="bar-chart" width="470" height="400" cssclasses="" plugins="" data="[object Object]" options="[object Object]"></mdb-bar-chart-stub>
            </div>
          </mdb-card-body-stub>
        </mdb-card-stub>
      </mdb-col-stub>
    </mdb-row-stub>
  </mdb-container-stub>
  <mdb-card-stub tag="div" class="m-3">
    <mdb-card-header-stub tag="div" color="indigo lighten-1">Data Table</mdb-card-header-stub>
    <datatable-stub salesdata="[object Object]"></datatable-stub>
  </mdb-card-stub>
</div>
`;
